File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if(preorder.length == 0 || inorder.length == 0){
            return null;
        }
        int rootIdx = 0;
        TreeNode root = new TreeNode(preorder[0]);

        //find the root in inorder array
        for(int i = 0; i < inorder.length; i++){
            if(inorder[i] == root.val){
                rootIdx = i;
            }
        }

        int[] preLeft = Arrays.copyOfRange(preorder, 1, rootIdx + 1);
        int [] preRight = Arrays.copyOfRange(preorder, rootIdx + 1, preorder.length);
        int[] inLeft = Arrays.copyOfRange(inorder, 0, rootIdx);
        int[] inRight = Arrays.copyOfRange(inorder, rootIdx + 1, inorder.length);

        root.left = buildTree(preLeft, inLeft);
        root.right = buildTree(preRight, inRight);

        return root;
    }
}

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    hashMap = dict()
    idx = 0
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        self.hashMap = dict()
        idx = 0
        
        for index, value in enumerate(inorder):
            self.hashMap[value] = index

        return self.helper(preorder, inorder, 0, len(inorder) - 1)
    
    def helper(self, preorder, inorder, start, end):
        #base
        if start > end:
            return None
        #logic
        root = TreeNode(preorder[self.idx])
        rootIdx = self.hashMap.get(preorder[self.idx])
        self.idx += 1

        root.left = self.helper(preorder, inorder, start, rootIdx - 1)
        root.right = self.helper(preorder, inorder, rootIdx + 1, end)
        return root